File: sdk/python/orbit_sdk/grants.py
from __future__ import annotations

import base64
import hashlib
import re
from dataclasses import dataclass
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import ec, utils

from .errors import INVALID_RESPONSE, error
from .jsonutil import fields, strict_int, text, unique_json

_CLAIM_TYPES: dict[str, Any] = {
    "iss": str,
    "aud": str,
    "sub": str,
    "jti": str,
    "iat": int,
    "nbf": int,
    "exp": int,
    "application_id": str,
    "environment_id": str,
    "activation_id": str,
    "installation_id": str,
    "binding_mode": str,
    "fingerprint": (str, type(None)),
    "fingerprint_provider": (str, type(None)),
    "policy_version": int,
    "entitlements": dict,
    "refresh_after": int,
    "offline_allowed": bool,
    "licence_expires_at": (int, type(None)),
}
_TOKEN_SEGMENTS = re.compile(r"^[A-Za-z0-9_-]+$")


def _invalid() -> None:
    raise error(INVALID_RESPONSE, "invalid_grant")


def _canonical_b64(value: str) -> bytes:
    if not isinstance(value, str) or not value or "=" in value or not _TOKEN_SEGMENTS.fullmatch(value):
        _invalid()
    try:
        padded = value + "=" * ((4 - len(value) % 4) % 4)
        decoded = base64.urlsafe_b64decode(padded.encode("ascii"))
    except (ValueError, UnicodeError) as exc:
        raise error(INVALID_RESPONSE, "invalid_grant") from exc
    if base64.urlsafe_b64encode(decoded).rstrip(b"=").decode("ascii") != value:
        _invalid()
    return decoded


def parse_header(token: str) -> tuple[dict[str, str], bytes, bytes, bytes]:
    if not isinstance(token, str) or len(token.encode("utf-8", "strict")) > 16384 or not token.isascii():
        _invalid()
    parts = token.split(".")
    if len(parts) != 3:
        _invalid()
    header_bytes, payload_bytes, signature = (_canonical_b64(part) for part in parts)
    if len(signature) != 64:
        _invalid()
    header = fields(unique_json(header_bytes), {"alg": str, "typ": str, "kid": str}, exact=True)
    if header["alg"] != "ES256" or header["typ"] != "orbit-access+jwt" or not text(header["kid"], minimum=1, maximum=128, ascii_only=True):
        _invalid()
    return header, header_bytes, payload_bytes, signature


class Keys:
    """A bounded set of validated P-256 verification keys."""

    def __init__(self, entries: dict[str, ec.EllipticCurvePublicKey]) -> None:
        self._entries = entries

    @classmethod
    def parse(cls, value: Any) -> Keys:
        if isinstance(value, (bytes, bytearray, str)):
            value = unique_json(value)
        if not isinstance(value, dict) or not isinstance(value.get("keys"), list) or not 1 <= len(value["keys"]) <= 8:
            _invalid()
        entries: dict[str, ec.EllipticCurvePublicKey] = {}
        for raw in value["keys"]:
            item = fields(raw, {"kty": str, "crv": str, "alg": str, "use": str, "kid": str, "x": str, "y": str}, exact=True)
            if item["kty"] != "EC" or item["crv"] != "P-256" or item["alg"] != "ES256" or item["use"] != "sig":
                _invalid()
            kid = item["kid"]
            if not text(kid, minimum=1, maximum=128, ascii_only=True) or kid in entries:
                _invalid()
            x, y = _canonical_b64(item["x"]), _canonical_b64(item["y"])
            if len(x) != 32 or len(y) != 32:
                _invalid()
            try:
                public = ec.EllipticCurvePublicNumbers(int.from_bytes(x, "big"), int.from_bytes(y, "big"), ec.SECP256R1()).public_key()
            except ValueError as exc:
                raise error(INVALID_RESPONSE, "invalid_jwks") from exc
            entries[kid] = public
        return cls(entries)

    def contains(self, token: str) -> bool:
        header, *_ = parse_header(token)
        return header["kid"] in self._entries


@dataclass(frozen=True)
class Expected:
    issuer: str
    application: str
    environment: str
    licence: str | None
    activation: str
    installation: str
    fingerprint: str | None
    fingerprint_provider: str | None
    credential_expires_at: int
    licence_expires_at: int | None
    now: int


def valid_entitlements(value: Any) -> bool:
    if not isinstance(value, dict) or len(value) > 64:
        return False
    for name, enabled in value.items():
        if not text(name, minimum=1, maximum=64, ascii_only=True) or name[0] not in "abcdefghijklmnopqrstuvwxyz":
            return False
        if any(char not in "abcdefghijklmnopqrstuvwxyz0123456789_" for char in name) or not isinstance(enabled, bool):
            return False
    return True


def verify(token: str, keys: Keys, expected: Expected) -> dict[str, Any]:
    header, header_bytes, payload, signature = parse_header(token)
    public = keys._entries.get(header["kid"])
    if public is None:
        _invalid()
    claims = fields(
        unique_json(payload),
        _CLAIM_TYPES,
        optional=("fingerprint", "fingerprint_provider", "licence_expires_at"),
    )
    if any(not strict_int(claims[name]) for name in ("iat", "nbf", "exp", "refresh_after", "policy_version")):
        _invalid()
    if claims["policy_version"] > (1 << 31) - 1 or claims["policy_version"] < 1:
        _invalid()
    if claims["licence_expires_at"] is not None and not strict_int(claims["licence_expires_at"]):
        _invalid()
    if not valid_entitlements(claims["entitlements"]):
        _invalid()
    for name in ("iss", "aud", "sub", "jti", "application_id", "environment_id", "activation_id", "installation_id", "binding_mode"):
        if not text(claims[name], maximum=128):
            _invalid()
    for name in ("fingerprint", "fingerprint_provider"):
        if claims[name] is not None and not text(claims[name], maximum=128):
            _invalid()
    signing_input = token.rsplit(".", 1)[0].encode("ascii")
    der_signature = utils.encode_dss_signature(int.from_bytes(signature[:32], "big"), int.from_bytes(signature[32:], "big"))
    try:
        public.verify(der_signature, signing_input, ec.ECDSA(hashes.SHA256()))
    except InvalidSignature as exc:
        raise error(INVALID_RESPONSE, "invalid_grant") from exc
    except (ValueError, TypeError) as exc:
        raise error(INVALID_RESPONSE, "invalid_grant") from exc
    audience = f"orbit:{expected.application}:{expected.environment}"
    bound = (
        claims["binding_mode"] == "none"
        and claims["fingerprint"] is None
        and claims["fingerprint_provider"] is None
        and expected.fingerprint is None
        and expected.fingerprint_provider is None
    ) or (
        claims["binding_mode"] == "hwid"
        and claims["fingerprint"] == expected.fingerprint
        and claims["fingerprint_provider"] == expected.fingerprint_provider
        and expected.fingerprint is not None
        and expected.fingerprint_provider is not None
    )
    allowance = 86400 if claims["offline_allowed"] else 300
    iat, nbf, exp = claims["iat"], claims["nbf"], claims["exp"]
    sat_add = lambda left, right: max(-(1 << 63), min((1 << 63) - 1, left + right))
    if (
        claims["iss"] != expected.issuer
        or claims["aud"] != audience
        or not claims["sub"] or len(claims["sub"].encode()) > 128
        or expected.licence is not None and claims["sub"] != expected.licence
        or not claims["jti"] or len(claims["jti"].encode()) > 128
        or claims["application_id"] != expected.application
        or claims["environment_id"] != expected.environment
        or claims["activation_id"] != expected.activation
        or claims["installation_id"] != expected.installation
        or not bound
        or iat < 0 or nbf != iat
        or iat > sat_add(expected.now, 30) or iat < sat_add(expected.now, -30)
        or exp <= expected.now or exp <= iat or exp > sat_add(iat, allowance)
        or exp > expected.credential_expires_at
        or claims["licence_expires_at"] != expected.licence_expires_at
        or claims["licence_expires_at"] is not None and exp > claims["licence_expires_at"]
        or claims["refresh_after"] <= iat or claims["refresh_after"] > exp
        or claims["refresh_after"] > sat_add(iat, 75)
        or claims["refresh_after"] < sat_add(iat, 45) and claims["refresh_after"] != exp
    ):
        _invalid()
    return claims

File: sdk/rust/src/grants.rs
use crate::{Error, Result};
use base64::{Engine, engine::general_purpose::URL_SAFE_NO_PAD};
use jsonwebtoken::{Algorithm, DecodingKey, Validation, decode};
use serde::{Deserialize, de};
use std::{collections::BTreeMap, fmt};

#[cfg(test)]
#[path = "grant_vectors.rs"]
mod vectors;

#[cfg(test)]
#[path = "grant_mutations.rs"]
mod mutations;

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Header {
    alg: String,
    typ: String,
    kid: String,
}
#[derive(Clone, Deserialize)]
pub struct Claims {
    pub iss: String,
    pub aud: String,
    pub sub: String,
    pub jti: String,
    pub iat: i64,
    pub nbf: i64,
    pub exp: i64,
    pub application_id: String,
    pub environment_id: String,
    pub activation_id: String,
    pub installation_id: String,
    pub binding_mode: String,
    pub fingerprint: Option<String>,
    pub fingerprint_provider: Option<String>,
    pub policy_version: i32,
    #[serde(deserialize_with = "entitlements")]
    pub entitlements: BTreeMap<String, bool>,
    pub refresh_after: i64,
    pub offline_allowed: bool,
    pub licence_expires_at: Option<i64>,
}
fn entitlements<'de, D: serde::Deserializer<'de>>(
    deserializer: D,
) -> std::result::Result<BTreeMap<String, bool>, D::Error> {
    struct Entries;
    impl<'de> de::Visitor<'de> for Entries {
        type Value = BTreeMap<String, bool>;
        fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            f.write_str("bounded unique boolean entitlements")
        }
        fn visit_map<A: de::MapAccess<'de>>(
            self,
            mut entries: A,
        ) -> std::result::Result<Self::Value, A::Error> {
            let mut map = BTreeMap::new();
            while let Some((key, value)) = entries.next_entry::<String, bool>()? {
                if key.is_empty()
                    || key.len() > 64
                    || !key.as_bytes()[0].is_ascii_lowercase()
                    || !key
                        .bytes()
                        .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == b'_')
                    || map.len() == 64
                    || map.insert(key, value).is_some()
                {
                    return Err(de::Error::custom("invalid entitlements"));
                }
            }
            Ok(map)
        }
    }
    deserializer.deserialize_map(Entries)
}
#[derive(Clone, Deserialize, serde::Serialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct Jwk {
    kty: String,
    crv: String,
    alg: String,
    #[serde(rename = "use")]
    purpose: String,
    kid: String,
    x: String,
    y: String,
}
#[derive(Clone, Deserialize, serde::Serialize)]
#[serde(deny_unknown_fields)]
pub(crate) struct Jwks {
    pub(crate) keys: Vec<Jwk>,
}
#[derive(Default)]
pub struct Keys(BTreeMap<String, DecodingKey>, BTreeMap<String, Jwk>);
impl Keys {
    pub fn parse(value: serde_json::Value) -> Result<Self> {
        let jwks: Jwks = serde_json::from_value(value).map_err(|_| Error::InvalidResponse)?;
        if jwks.keys.is_empty() || jwks.keys.len() > 8 {
            return Err(Error::InvalidResponse);
        }
        let mut keys = BTreeMap::new();
        let mut public_keys = BTreeMap::new();
        for key in jwks.keys {
            if key.kty != "EC"
                || key.crv != "P-256"
                || key.alg != "ES256"
                || key.purpose != "sig"
                || key.kid.is_empty()
                || key.kid.len() > 128
                || !key.kid.is_ascii()
                || canonical(&key.x)?.len() != 32
                || canonical(&key.y)?.len() != 32
            {
                return Err(Error::InvalidResponse);
            }
            let public = DecodingKey::from_ec_components(&key.x, &key.y)
                .map_err(|_| Error::InvalidResponse)?;
            public_keys.insert(key.kid.clone(), key.clone());
            if keys.insert(key.kid, public).is_some() {
                return Err(Error::InvalidResponse);
            }
        }
        Ok(Self(keys, public_keys))
    }
    pub(crate) fn public_key(&self, token: &str) -> Result<serde_json::Value> {
        let key = self
            .1
            .get(&header(token)?.kid)
            .ok_or(Error::InvalidResponse)?;
        serde_json::to_value(Jwks {
            keys: vec![key.clone()],
        })
        .map_err(|_| Error::InvalidResponse)
    }
    pub fn contains(&self, token: &str) -> Result<bool> {
        Ok(self.0.contains_key(&header(token)?.kid))
    }
}
fn canonical(value: &str) -> Result<Vec<u8>> {
    let bytes = URL_SAFE_NO_PAD
        .decode(value)
        .map_err(|_| Error::InvalidResponse)?;
    if URL_SAFE_NO_PAD.encode(&bytes) != value {
        return Err(Error::InvalidResponse);
    }
    Ok(bytes)
}
fn header(token: &str) -> Result<Header> {
    if token.len() > 16384 {
        return Err(Error::InvalidResponse);
    }
    let parts: Vec<_> = token.split('.').collect();
    if parts.len() != 3 || canonical(parts[2])?.len() != 64 {
        return Err(Error::InvalidResponse);
    }
    let header: Header =
        serde_json::from_slice(&canonical(parts[0])?).map_err(|_| Error::InvalidResponse)?;
    canonical(parts[1])?;
    if header.alg != "ES256" || header.typ != "orbit-access+jwt" || header.kid.len() > 128 {
        return Err(Error::InvalidResponse);
    }
    Ok(header)
}
pub struct Expected<'a> {
    pub issuer: &'a str,
    pub application: &'a str,
    pub environment: &'a str,
    pub licence: Option<&'a str>,
    pub activation: &'a str,
    pub installation: &'a str,
    pub fingerprint: Option<&'a str>,
    pub fingerprint_provider: Option<&'a str>,
    pub credential_expires_at: Option<i64>,
    pub licence_expires_at: Option<i64>,
    pub now: i64,
}
pub fn verify(token: &str, keys: &Keys, expected: &Expected<'_>) -> Result<Claims> {
    jsonwebtoken::crypto::aws_lc::DEFAULT_PROVIDER
        .install_default()
        .ok();
    let header = header(token)?;
    let key = keys.0.get(&header.kid).ok_or(Error::InvalidResponse)?;
    let audience = format!("orbit:{}:{}", expected.application, expected.environment);
    let mut validation = Validation::new(Algorithm::ES256);
    validation.validate_exp = false;
    validation.validate_nbf = false;
    validation.leeway = 0;
    validation.set_audience(&[&audience]);
    validation.set_issuer(&[expected.issuer]);
    validation.set_required_spec_claims(&["exp", "nbf", "iat", "iss", "aud", "sub"]);
    let claims: Claims = decode::<Claims>(token, key, &validation)
        .map_err(|_| Error::InvalidResponse)?
        .claims;
    let allowance = if claims.offline_allowed { 86400 } else { 300 };
    let (refresh_min, refresh_max) =
        if expected.credential_expires_at.is_none() && claims.offline_allowed {
            (675, 1125)
        } else {
            (45, 75)
        };
    let bound = match (expected.fingerprint, expected.fingerprint_provider) {
        (None, None) => {
            claims.binding_mode == "none"
                && claims.fingerprint.is_none()
                && claims.fingerprint_provider.is_none()
        }
        (Some(fingerprint), Some(provider)) => {
            claims.binding_mode == "hwid"
                && claims.fingerprint.as_deref() == Some(fingerprint)
                && claims.fingerprint_provider.as_deref() == Some(provider)
        }
        _ => false,
    };
    if claims.iss != expected.issuer
        || claims.aud != audience
        || claims.sub.is_empty()
        || claims.sub.len() > 128
        || expected.licence.is_some_and(|id| id != claims.sub)
        || claims.jti.is_empty()
        || claims.jti.len() > 128
        || claims.application_id != expected.application
        || claims.environment_id != expected.environment
        || claims.activation_id != expected.activation
        || claims.installation_id != expected.installation
        || !bound
        || claims.policy_version < 1
        || claims.iat < 0
        || claims.nbf != claims.iat
        || claims.iat > expected.now.saturating_add(30)
        || claims.iat < expected.now.saturating_sub(30)
        || claims.exp <= expected.now
        || claims.exp <= claims.iat
        || claims.exp > claims.iat.saturating_add(allowance)
        || expected
            .credential_expires_at
            .is_some_and(|expiry| claims.exp > expiry)
        || claims.licence_expires_at != expected.licence_expires_at
        || claims
            .licence_expires_at
            .is_some_and(|expiry| claims.exp > expiry)
        || claims.refresh_after <= claims.iat
        || claims.refresh_after > claims.exp
        || claims.refresh_after > claims.iat.saturating_add(refresh_max)
        || (claims.refresh_after < claims.iat.saturating_add(refresh_min)
            && claims.refresh_after != claims.exp)
    {
        return Err(Error::InvalidResponse);
    }
    Ok(claims)
}

#[cfg(test)]
mod tests {
    use super::*;
    use jsonwebtoken::{EncodingKey, Header};
    use serde::Serialize;
    const NOW: i64 = 1_800_000_000;
    pub(super) const PRIVATE: &[u8] = include_bytes!("../tests/fixtures/es256-test-private.pem");
    pub(super) fn claims() -> serde_json::Value {
        serde_json::json!({"iss":"https://orbit.example.test","aud":"orbit:app:test","sub":"licence","jti":"random",
            "iat":NOW,"nbf":NOW,"exp":NOW+300,"application_id":"app","environment_id":"test","activation_id":"activation",
            "installation_id":"installation","binding_mode":"none","policy_version":1,"entitlements":{"export":true},"refresh_after":NOW+60,"offline_allowed":false})
    }
    fn keys() -> Keys {
        Keys(
            BTreeMap::from([(
                "test-key".into(),
                DecodingKey::from_ec_pem(include_bytes!("../tests/fixtures/es256-test-public.pem"))
                    .unwrap(),
            )]),
            BTreeMap::new(),
        )
    }
    pub(super) fn expected() -> Expected<'static> {
        Expected {
            issuer: "https://orbit.example.test",
            application: "app",
            environment: "test",
            licence: Some("licence"),
            activation: "activation",
            installation: "installation",
            fingerprint: None,
            fingerprint_provider: None,
            credential_expires_at: Some(NOW + 3600),
            licence_expires_at: None,
            now: NOW,
        }
    }
    pub(super) fn sign<T: Serialize>(value: &T, header: Option<Header>) -> String {
        jsonwebtoken::crypto::aws_lc::DEFAULT_PROVIDER
            .install_default()
            .ok();
        let mut default = Header::new(Algorithm::ES256);
        default.typ = Some("orbit-access+jwt".into());
        default.kid = Some("test-key".into());
        jsonwebtoken::encode(
            &header.unwrap_or(default),
            value,
            &EncodingKey::from_ec_pem(PRIVATE).unwrap(),
        )
        .unwrap()
    }
    #[test]
    fn valid_grant_and_policy_caps() {
        let verified = verify(&sign(&claims(), None), &keys(), &expected()).unwrap();
        assert!(verified.entitlements["export"]);
        let mut offline = claims();
        offline["offline_allowed"] = true.into();
        offline["exp"] = (NOW + 900).into();
        assert!(verify(&sign(&offline, None), &keys(), &expected()).is_ok());
        let mut timed = expected();
        timed.licence_expires_at = Some(NOW + 250);
        assert!(verify(&sign(&claims(), None), &keys(), &timed).is_err());
    }
    #[test]
    fn scope_types_lifetime_and_missing_claims_fail_closed() {
        for (name, value) in [
            ("iss", json!("https://evil.test")),
            ("aud", json!("orbit:app:live")),
            ("sub", json!("other")),
            ("application_id", json!("other")),
            ("environment_id", json!("live")),
            ("activation_id", json!("other")),
            ("installation_id", json!("other")),
            ("binding_mode", json!("unknown")),
            ("policy_version", json!(0)),
            ("offline_allowed", json!("true")),
            ("exp", json!(NOW)),
            ("exp", json!(NOW + 301)),
            ("iat", json!(NOW + 31)),
            ("nbf", json!(NOW + 1)),
            ("refresh_after", json!(NOW + 76)),
            ("refresh_after", json!(NOW + 44)),
            ("entitlements", json!({"Export":true})),
            ("entitlements", json!({"export":"true"})),
            ("licence_expires_at", json!(NOW + 200)),
        ] {
            let mut value2 = claims();
            value2[name] = value;
            assert!(
                verify(&sign(&value2, None), &keys(), &expected()).is_err(),
                "{name}"
            );
        }
        for field in [
            "iss",
            "aud",
            "sub",
            "jti",
            "iat",
            "nbf",
            "exp",
            "application_id",
            "environment_id",
            "activation_id",
            "installation_id",
            "binding_mode",
            "policy_version",
            "entitlements",
            "refresh_after",
            "offline_allowed",
        ] {
            let mut value = claims();
            value.as_object_mut().unwrap().remove(field);
            assert!(
                verify(&sign(&value, None), &keys(), &expected()).is_err(),
                "{field}"
            );
        }
    }
    use serde_json::json;
    #[test]
    fn bad_signatures_untrusted_headers_duplicates_and_algorithms() {
        let token = sign(&claims(), None);
        let mut changed = token.as_bytes().to_vec();
        let start = token.rfind('.').unwrap() + 1;
        changed[start] = if changed[start] == b'A' { b'B' } else { b'A' };
        assert!(verify(std::str::from_utf8(&changed).unwrap(), &keys(), &expected()).is_err());
        for kind in ["kid", "typ", "crit", "jku"] {
            let mut h = Header::new(Algorithm::ES256);
            h.typ = Some("orbit-access+jwt".into());
            h.kid = Some("test-key".into());
            match kind {
                "kid" => h.kid = Some("live-key".into()),
                "typ" => h.typ = Some("JWT".into()),
                "crit" => h.crit = Some(vec!["unknown".into()]),
                _ => h.jku = Some("https://evil.test".into()),
            }
            assert!(verify(&sign(&claims(), Some(h)), &keys(), &expected()).is_err());
        }
        for token in ["", "a.b.c", "a.b.c.d", "eyJhbGciOiJub25lIn0.e30."] {
            assert!(verify(token, &keys(), &expected()).is_err());
        }
        #[derive(Clone)]
        struct Duplicate;
        impl Serialize for Duplicate {
            fn serialize<S: serde::Serializer>(
                &self,
                serializer: S,
            ) -> std::result::Result<S::Ok, S::Error> {
                use serde::ser::SerializeMap;
                let mut map = serializer.serialize_map(None)?;
                for (key, value) in claims().as_object().unwrap() {
                    map.serialize_entry(key, value)?;
                }
                map.serialize_entry("exp", &(NOW + 300))?;
                map.end()
            }
        }
        assert!(verify(&sign(&Duplicate, None), &keys(), &expected()).is_err());
        let duplicate = serde_json::to_string(&claims())
            .unwrap()
            .replace("\"export\":true", "\"export\":true,\"export\":false");
        assert!(serde_json::from_str::<Claims>(&duplicate).is_err());
        assert!(
            serde_json::from_str::<Header>(
                r#"{"alg":"ES256","alg":"HS256","typ":"orbit-access+jwt","kid":"test-key"}"#
            )
            .is_err()
        );
    }
}
